Design.md

```diff
@@ -309,3 +332,29 @@ Get the platform usage of global (without query parameter), tenant (with tenant_
         * `key` : `offset` and `value` : `marker`
     1. `AssumeRoleCache` for all the `User` and `S3 Credential` APIs
         * `key` : `RoleArn` and `value` : `Credentials`
+
+## SecretKey Encryption Strategy
+
+1. For encrypting/ decrypting the `secret-keys` a `cipher` algorithm, and a `secret-password` will be used. These properties will be provided in `crypto.properties` config file.
+1. A `key-id` property will be provided for each `secret-password` in the `crypto.properties` config file to maintain the version of the key and will be used in the Key rotation.
+1. Meta-information will be encoded with base64 and added at the beginning of the encrypted string. This meta information will include the `key-id` of the `secret-password` used.
+1. All the secret-passwords in the `crypto.properties` config file will be base64 encoded. These secret-passwords will be decoded using base64 before encrypting the `secret-keys`
+1. Initially only `AES256GCM` cipher algorithm will be supported. (`secret-box` algorithm is suggested to be implemented for future releases)
+1. Adaptability to change the cipher algorithm will be provided.
+1. Example `crypto.properties`
+    ```yaml
+    keys:
+      - id: 2
+        cipher: Secretbox
+        secretKey: dGhpc2lzYXJlYWxseWxvbmdhbmRzdHJvbmdrZXk=
+    ```
+1. **Key Rotation:**
+    * Key rotation is used to replace a key with another one once the original key is compromised or too old to use.
+    * A new API will be provided with the `OSIS` application to do the key rotation of all the encrypted `secret-key` data on Redis database. 
```

Comment:
I'm not convinced this must be an API. Please elaborate as to why this must necessarily be API-driven.

Reply:
As we are going to provide these keys in a properties file, we definitely need to reload this file on Spring boot.
Easiest and cleanest way to do this is using Spring Actuator `/refresh` API. 

I thought of two approaches to do the key rotation:
1. As we are going to call the `/refresh` API anyways, it is easy to reuse the same code used for encryption/decryption to refresh the cipher texts on Redis during this refresh API call.
  * By calling this API, the program will retry and will not return until all the keys are rotated.
2. We can also provide a small standalone application jar with the same encryption/decryption code packed in it. 
  * First this standalone application can CURL `/refresh` API to refresh the properties file and OSIS can use latest keys for new usage.
  * Then this program will refresh the cipher texts on Redis using the latest key. 
  * If that is not a success we can recall it

Please suggest which one soubds better.

Reply:
I prefer the latter. The actual rotation is not the driver's job: the driver (assuming it's correctly configured) works just fine with all secrets, whether they're encrypted using some old key, some new key, whether they're being rotated,... The actual rotation can be a process that's running in parallel with the driver, and hence shouldn't (IMHO) be 'inside' the driver. Having this as an out-of-band operation one can invoke (essentially) at any point in time enforces us to keep things 'clean' and not have the driver to more than it should.

Reply:
Also, is such `/refresh` API call really needed? Can't the driver be restarted when there's a new config? What would the effect of such (temporal) unavailability of the driver be?

Keep in mind key rotation is not something one does every other day. It's a very 'special' operation.

Reply:
Updated in 1fce348

Reply:
Related to `/refresh` API, my only concern is after the key rotation application run is complete, if a new user is created, until OSIS access the existing values on database it will not refresh the config and will not know about the update in the keys.

There is a possibility that new entries might still be created on database using the old keys after the key rotation. My suggestion would be:
1. Do not delete the old key form the original config file. Add only the new key above it with the incremented `key-id`.
2. After data is retrieved from database, `OSIS` application should always check if the `key-id` found is the **latest** `key-id`. If not re-encrypt the values on database using the **latest** `key-id`